### ISSUE TITLE
sout-keep doesn't keep the stream open when changing audio tracks with an AudioListPlayerComponent

### ISSUE THREAD

**Issue by BananaPuncher714** [caprica/vlcj#1008]
I'm streaming a list of arbitrary files to an Icecast server, but whenever a track finishes playing, the client stops the stream even though there are more tracks being played afterwards. I have `--sout-keep` added to the factory options, and `:sout-keep` in the playing options, but those do not appear to do anything.

It works natively, if I execute this on the command line: `vlc -vvv playlist :sout=#transcode{vcodec=none,acodec=mp3,ab=320,channels=2,samplerate=44100,scodec=none}:std{access=shout{name="Icecast Test"},mux=mp3,dst="//source:hackme@localhost:8000/radio"} :no-sout-all :sout-keep --sout-shout-mp3` where `playlist` is a folder containing multiple audio tracks. Whenever vlc changes tracks, the player doesn't stop the playback.

This is the code I am using for testing, where I add the mrl of the folder containing the music:
```java
import java.io.File;

import uk.co.caprica.vlcj.factory.MediaPlayerFactory;
import uk.co.caprica.vlcj.player.base.MediaPlayer;
import uk.co.caprica.vlcj.player.base.MediaPlayerEventAdapter;
import uk.co.caprica.vlcj.player.component.AudioListPlayerComponent;

public class SoutKeepTest {
	private static File BASE = new File( System.getProperty( "user.dir" ) );

	private static final String[] FACTORY_OPTIONS = {
			"--no-sout-all",
			"--sout-shout-mp3",
			"--sout-keep"
	};
	private static final String[] OPTIONS = {
			":sout=#transcode{vcodec=none,acodec=mp3,ab=320,channels=2,samplerate=44100,scodec=none}:std{access=shout{name='Icecast Test'},mux=mp3,dst=//source:hackme@localhost:8000/radio}",
			":sout-keep"
	};
	
	public static void main( String[] args ) {
		MediaPlayerFactory factory = new MediaPlayerFactory( FACTORY_OPTIONS );
		AudioListPlayerComponent component = new AudioListPlayerComponent( factory );
		component.mediaPlayer().events().addMediaPlayerEventListener( new MediaPlayerEventAdapter() {
		    @Override
		    public void playing( MediaPlayer mediaPlayer ) {
		    	System.out.println( "Started playing mediaPlayer" );
		    }

		    @Override
		    public void finished( MediaPlayer mediaPlayer ) {
		    	System.out.println( "Finished playing" );
		    }

		    @Override
		    public void error( MediaPlayer mediaPlayer ) {
		    	System.out.println( "Stopped playing forcefully" );
		    }
		} );
		
		component.mediaListPlayer().list().media().add( new File( BASE, "playlist" ).getAbsolutePath(), OPTIONS );

		System.out.println( "Starting..." );
		component.mediaListPlayer().controls().play();
		
		while ( true ) {
			try {
				Thread.sleep( 1000 );
			} catch ( InterruptedException e ) {
				e.printStackTrace();
			}
		}
	}
}
```

The music streams fine, and all the tracks get played like expected, but the stream is closed clientside. Everything is done on the same machine.

I'm using these versions of software:
- vlcj 4.6.0
- vlc 3.0.11 Vetinari
- java 11.0.2
- Icecast 2.4.4
- Foobar2000 1.5.5 for listening to the stream
- 64 bit Windows 10 Pro version 1909

**Comment by caprica** [comment]
On the face of it I can't see anything wrong. I can have a closer look later.

That's a really nice report btw.

**Comment by caprica** [comment]
Just to be clear, the MRL you add is a directory name?

**Comment by caprica** [comment]
The reason I ask about that...

My impression was if you added a directory, what you would actually get is one media item in your media list (the directory) - and that one media item would have sub-items for each file in that directory. That's sub-items, not top-level items.

And what vlcj does for you by default is to the first sub-item of a media. To be fair, I'd expect it to play *all* the sub-items but I don't know for sure what it will do.

So I think what you might need instead is to add the actual files individually to your media list and play those, rather than adding the directory.

It is just a guess though.

**Comment by caprica** [comment]
Oh sorry, I think I may have misunderstood, you're saying it DOES play all the tracks correctly? But after the last track is closes the client connection?

**Comment by BananaPuncher714** [comment]
Sorry, I probably wasn't too clear with the behavior of the code I posted.

I added the MRL of a directory to the AudioListPlayerComponent's `ListAPI`. When I play the component through the `ControlsAPI`, the program plays each audio track in the directory, one after another. This behavior is the exact same as if I had provided the directory to vlc through the command line and works as expected.

However, when listening to the AudioListPlayerComponent stream through the client(Foobar200), each time the code finishes playing an audio track, the client detects the end of stream and stops playing. I have to manually start the stream again each time. This does not occur when I run vlc through the windows command prompt; the client simply treats the multiple audio tracks as a single stream and continues playing without stopping. I have also used vlc to listen to both streams, and experienced the same behavior.

I have tried adding each file separately, but the problem still persists. Here is the code snippet:
```java
		for ( File file : new File( BASE, "playlist" ).listFiles() ) {
			System.out.println( "Added file " + file + " to the playlist" );
			component.mediaListPlayer().list().media().add( file.getAbsolutePath(), OPTIONS );
		}
```

I can replicate the code's behavior with vlc from the command line if I omit the `:sout-keep` option, which leads me to believe that may be the cause, somehow.

**Comment by caprica** [comment]
OK, that's more clear.

**Comment by caprica** [comment]
Can you print the return value of ".media().add(...)."

**Comment by BananaPuncher714** [comment]
Adding the directory and each file separately returned true

**Comment by caprica** [comment]
I'm not all that familiar with what all these sout options are for, but in my own code where I stream I have this along with sout-keep:

mediaPlayer.media().play(media,
            ":no-sout-rtp-sap",
            ":no-sout-standard-sap",
            ":sout-keep"
        );

Maybe those session announcement protocol announcements are confusing the client.

I really have no other clue here I'm afraid.

Did you try "-vvv" on your factory and checking the native logs to see if you can see any hints - like confirming that sout-keep was actually used?

**Comment by caprica** [comment]
But then you don't use those when you play with VLC itself do you.

Hmm...

**Comment by BananaPuncher714** [comment]
No, I will try those options and see what I get.
- `no-sout-all` prevents elementary streams from getting sent, which I don't think makes much difference. I've had it enabled and disabled without noticeable effects.
- `sout-shout-mp3` seems to allow vlc to stream in mp3 format, instead of ogg to Icecast, which I use for both the code and on the command line.
- `sout-keep` states that it *should* keep the stream open instead of closing it when a track is finished playing

The transcode option is just for streaming to Icecast.

**Comment by caprica** [comment]
What I would normally do in cases like this, when dealing with these non-API kind of things, is to create my factory with "-vvv" and grab the logs.

Then run "vlc" with "-vvv" and grab those logs.

Then painstakingly compare the two.

Apart from that I have no more suggestions, sorry.

**Comment by caprica** [comment]
I don't have the same setup with you with Icecast but I hacked your test class to stream over HTTP so I could give it a quick test myself.

I admit I'm not exactly comparing apples with apples here, but it's the best I can do right now to approximate your issue.

I used the native VLC application as the client.

There was a brief glitch at playback start, but after that the first track played through just fine. I also noted the time widget in VLC updated as the stream played.

In the "server" I see the first track end and the second begin playing:

```
Starting...
Started playing mediaPlayer                     <--- starts playing the "directory"
Finished playing                                <--- finishes playing the directory
Started playing mediaPlayer                     <--- immediately starts playing the first track
Finished playing                                <--- first track ends
[in your case, the client disconnects and stops playing here?]
Started playing mediaPlayer                     <--- second track begins
Finished playing                                <--- second track ends
Started playing mediaPlayer                     <--- third track begins
```
While the track is playing there's a glitch/drop-out but it does come back quite quickly. When it comes back the time widget starts updating again.

Same behaviour with the time widget, the second third track plays but the widget stops updating.

Now, fair enough there are some key differences - I am using simple HTTP streaming rather than your Icecast arrangement.

Again to be fair it did glitch on the fourth track and totally stop so I'm not going to pretend and say it was working perfectly.

BUT, despite a few glitches it appeared to be working properly and keeping the same sout stream alive across the playlist at least for a while.

So while the test results are not perfect, I think it was demonstrating that sout-keep was working.

I then did another test where I removed the sout-keep options, and indeed the VLC client stopped after the first track.

I'm also using Linux, not Windows.

The other difference I spotted between my case and yours is that anywhere in the sout string I see a ":" I put that as its own option - you have those concatenated. I'd be surprised if this is the issue, but it's worth a try.

I'm not sure where that leaves this issue.

Here's the hacked up code I used:

```
public class SoutKeepTest {
    private static File BASE = new File( System.getProperty( "user.dir" ) );

    private static final String[] FACTORY_OPTIONS = {
        "--no-sout-all",
        "--sout-shout-mp3",
        "--sout-keep"
    };
    private static final String[] OPTIONS = {
//        ":sout=#transcode{vcodec=none,acodec=mp3,ab=320,channels=2,samplerate=44100,scodec=none}:std{access=shout{name='Icecast Test'},mux=mp3,dst=//source:hackme@localhost:8000/radio}",
//        ":sout-keep"
        ":sout=#transcode{vcodec=none,acodec=mp3,ab=320,channels=2,samplerate=44100,scodec=none}",
        formatHttpStream("127.0.0.1", 5555),
        ":sout-keep"
    };

    public static void main( String[] args ) {
        MediaPlayerFactory factory = new MediaPlayerFactory( FACTORY_OPTIONS );
        AudioListPlayerComponent component = new AudioListPlayerComponent( factory );
        component.mediaPlayer().events().addMediaPlayerEventListener( new MediaPlayerEventAdapter() {
            @Override
            public void playing( MediaPlayer mediaPlayer ) {
                System.out.println( "Started playing mediaPlayer" );
            }

            @Override
            public void finished( MediaPlayer mediaPlayer ) {
                System.out.println( "Finished playing" );
            }

            @Override
            public void error( MediaPlayer mediaPlayer ) {
                System.out.println( "Stopped playing forcefully" );
            }
        } );

//        component.mediaListPlayer().list().media().add( new File( BASE, "playlist" ).getAbsolutePath(), OPTIONS );
        component.mediaListPlayer().list().media().add( new File( "/home/music//" ).getAbsolutePath(), OPTIONS );

        System.out.println( "Starting..." );
        component.mediaListPlayer().controls().play();

        while ( true ) {
            try {
                Thread.sleep( 1000 );
            } catch ( InterruptedException e ) {
                e.printStackTrace();
            }
        }
    }

    private static String formatHttpStream(String serverAddress, int serverPort) {
        StringBuilder sb = new StringBuilder(60);
        sb.append(":sout=#duplicate{dst=std{access=http,mux=ts,");
        sb.append("dst=");
        sb.append(serverAddress);
        sb.append(':');
        sb.append(serverPort);
        sb.append("}}");
        return sb.toString();
    }
}
```

**Comment by BananaPuncher714** [comment]
Ok, thanks. I'll give it a try. I've been looking over the logs like you suggested, but there doesn't seem to be any indication of which flags are included. If I omit `sout-keep`, it does stop playback without being able to restart, so you are definitely right about it working.

**Comment by BananaPuncher714** [comment]
I've tried the modified code you used, however, splitting the colons onto separate lines doesn't appear to work for Windows. I am forced to keep the `sout` modules all on one line, or else the audio stops playing completely. When I replaced the access with http, and listened to it with VLC, it still ended the playback after each track. I don't know where the issue lies either, but I don't think there's much that can be done about it. I'll try again on Linux in the future, but for now I will look for alternatives/try again every day until I find something that works.

Thanks for spending time on trying to troubleshoot, vlcj has been extremely useful for me.
